### ISSUE TITLE
Using RandomGenerator as a drop-in replacement for numpy's RandomState

### ISSUE THREAD

**Issue by praveenv253** [bashtage/randomgen#48]
This might be more of a feature request than an issue, per se. I was trying to use a `randomgen.generator.RandomGenerator` instance for repeated, stratified K-fold cross-validation using scikit-learn. I run several classification programs in parallel, and each uses cross-validation - I'd like each cross-validator to have a fixed seed, but to belong to an independent stream, which is why randomgen seemed to be a very attractive package.

The issue I ran into was that scikit-learn did not recognize a RandomGenerator object as a RandomState instance, and so it rejected it. Indeed, since RandomGenerator is a Cython class, I don't think it can inherit from RandomState.

I was looking for workarounds, and found that I could hack together something like this:

```python
import numpy as np
import randomgen
from sklearn.model_selection import RepeatedStratifiedKFold 

# Create a new class which subclasses RandomState
class MyRandomGenerator(np.random.RandomState): 
    def __init__(self, source=None): 
        if source is not None: 
            # Override all methods coming from RandomState using a RandomGenerator instance
            for d in source.__dir__(): 
                if not d.startswith('_'): 
                    self.__dict__.update({d: getattr(source, d)}) 

rg = randomgen.generator.RandomGenerator(randomgen.Philox(42))
myrg = MyRandomGenerator(rg)

rskf = RepeatedStratifiedKFold(n_splits=2, n_repeats=2, random_state=myrg) 
X = np.arange(20).reshape((10, 2))
y = (np.random.rand(10) < 0.5).astype(int)
print([(train, test) for train, test in rskf.split(X, y)])
print(myrg.shuffle.__self__ is rg)
```

I'm sure this won't work in general, but it works here, since I was only intent on preserving the `shuffle` method. Is there a better way to do this, and do you plan to have a built-in mechanism for something like this?

**Comment by bashtage** [comment]
The best fix would be for scikit-learn to duck type the source of randomness they use, rather than assume it must look one particular way.

Your solution is reasonable.  It is like how `LegacyGenerator` (python class) composes functions from `RandomGenerator` and `_LegacyGenerator` (cython class).

The built-in will come when randomgen moves into NumPy and replaces RandomState.  Then scikit-learn will update to support the new infrastrucutre, but perhaps they would accept a PR to do duck typing?

**Comment by praveenv253** [comment]
I don't disagree with you, but I had a peek at the scikit-learn source code, and they seem to be using a common `check_random_state` function for all functions that accept an rng parameter, which is where an `isinstance(random_state, np.random.RandomState)` check is made. Replacing this function to duck-type in every use-case, while also providing comprehensible error-messages, is most likely not an easy task, and is probably bad for the maintainability of scikit-learn.

Also, the `isinstance` check is not entirely unreasonable, and is almost a duck-test in and of itself. It just so happens that `RandomGenerator` could not be inherited from `RandomState` because it's a Cython class. Otherwise there would be no reason _not_ to, correct?

I guess that once randomgen is merged with numpy, things will fix themselves. Until then I'll rely on this workaround. Depending on how long it'll be until merger, do you think it would be of use to others to include a version of such a workaround in the randomgen codebase temporarily?

Thanks anyway!

**Comment by bashtage** [comment]
I could see the utility of a class like this.  Ideally it would inherit from RS and then (a) use RG instances where appropriate and (b) wrap state to look like a RS state (a function, not a prop, since some downstream projects might expect to `set/get_state`).

(b) is a little work, but should be pretty easy.

**Comment by bashtage** [comment]
@praveenv253 I'm going to close this since the effort to [integrate randomgen into numpy as started](https://github.com/numpy/numpy/pull/13163), and so I'm not going to make any large feature changes.  The good news is that once it is in numpy, I'm sure it will be adopted by the ecosystem.

**Comment by praveenv253** [comment]
That's great news! Thanks!
